extensions/druid-rocketmq/src/main/java/io/druid/firehose/rocketmq/RocketMQFirehoseFactory.java

```diff
@@ -0,0 +1,537 @@
+            public InputRow nextRow() {
+                for (Map.Entry<MessageQueue, ConcurrentSkipListSet<MessageExt>> entry : messageQueueTreeSetMap.entrySet()) {
+                    if (!entry.getValue().isEmpty()) {
+                        MessageExt message = entry.getValue().pollFirst();
+                        InputRow inputRow = theParser.parse(ByteBuffer.wrap(message.getBody()));
+
+                        if (!windows.keySet().contains(entry.getKey())) {
+                            windows.put(entry.getKey(), new ConcurrentSkipListSet<Long>());
+                        }
+                        windows.get(entry.getKey()).add(message.getQueueOffset());
+                        return inputRow;
+                    }
+                }
+
+                // should never happen.
+                return null;
+            }
+
+            @Override
+            public Runnable commit() {
+                return new Runnable() {
+                    @Override
+                    public void run() {
+                        OffsetStore offsetStore = defaultMQPullConsumer.getOffsetStore();
+                        Set<MessageQueue> updated = new HashSet<>();
+                        // calculate offsets according to consuming windows.
+                        for (ConcurrentHashMap.Entry<MessageQueue, ConcurrentSkipListSet<Long>> entry : windows.entrySet()) {
+                            while (!entry.getValue().isEmpty()) {
+
+                                long offset = offsetStore.readOffset(entry.getKey(), ReadOffsetType.MEMORY_FIRST_THEN_STORE);
+                                if (offset + 1 > entry.getValue().first()) {
+                                    entry.getValue().pollFirst();
+                                } else if (offset + 1 == entry.getValue().first()) {
+                                    entry.getValue().pollFirst();
+                                    offsetStore.updateOffset(entry.getKey(), offset + 1, true);
+                                    updated.add(entry.getKey());
+                                } else {
+                                    break;
+                                }
+
+                            }
+                        }
+                        offsetStore.persistAll(updated);
+                    }
+                };
+            }
+
+            @Override
+            public void close() throws IOException {
+                defaultMQPullConsumer.shutdown();
+                pullMessageService.shutdown(false);
+            }
+        };
+    }
+
+
+    /**
+     * Pull request.
+     */
+    class DruidPullRequest {
+        private MessageQueue messageQueue;
+        private String tag;
+        private long nextBeginOffset;
+        private int pullBatchSize;
+        private boolean longPull;
+        private CountDownLatch countDownLatch;
+        private PullResult pullResult;
+        private boolean successful;
+
+        public DruidPullRequest() {
```

Comment:
can we pass the variables in the constructor instead of using setters? let's keep mutable variables to a minimum